wgengine/magicsock: TestDiscokeyChange is flaky

### What is the issue?

Sample failure from https://github.com/tailscale/tailscale/commit/0303ec44c31e5adc96994168303cc75ab3c5b45e CI run:

```
--- FAIL: TestDiscokeyChange (10.26s)
    magicsock.go:2140: magicsock: SetPrivateKey called (init)
    magicsock.go:635: [v1] magicsock: starting endpoint update (set-private-key)
    wglog.go:58: [v2] UDP bind has been updated
    wglog.go:58: [v2] Interface state was Down, requested Up, now Up
    wglog.go:58: [v2] Routine: receive incoming receiveDERP - started
    magicsock.go:2643: [v1] magicsock: ReSTUN: endpoint update active, need another later ("derp-map-update")
    wglog.go:58: [v2] Routine: receive incoming v4 - started
    wglog.go:58: [v2] Routine: receive incoming v6 - started
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53541 derp=1 derpdist=
    magicsock.go:984: magicsock: home is now derp-1 (test)
    magicsock.go:2368: magicsock: endpoints changed: 127.0.0.1:53541 (stun)
    magicsock.go:1328: magicsock: adding connection to derp-1 for home-keep-alive
    magicsock.go:2356: magicsock: 1 active derp conns: derp-1=cr0s,wr0s
    magicsock.go:635: [v1] magicsock: starting endpoint update (derp-map-update)
    derphttp_client.go:287: derphttp.Client.Connect: connecting to derp-1 (test)
    magicsock.go:2140: magicsock: SetPrivateKey called (init)
    wglog.go:58: [v2] UDP bind has been updated
    wglog.go:58: [v2] Interface state was Down, requested Up, now Up
    wglog.go:58: [v2] Routine: receive incoming receiveDERP - started
    magicsock.go:635: [v1] magicsock: starting endpoint update (derp-map-update)
    magicsock.go:1536: magicsock: derp-1 connected; connGen=1
    magicsock.go:2643: [v1] magicsock: ReSTUN: endpoint update active, need another later ("set-private-key")
    wglog.go:58: [v2] Routine: receive incoming v4 - started
    wglog.go:58: [v2] Routine: receive incoming v6 - started
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53541 derp=1 derpdist=1v4:1ms
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53545 derp=1 derpdist=
    magicsock.go:984: magicsock: home is now derp-1 (test)
    magicsock.go:2368: magicsock: endpoints changed: 127.0.0.1:53545 (stun)
    magicsock.go:1328: magicsock: adding connection to derp-1 for home-keep-alive
    magicsock.go:2356: magicsock: 1 active derp conns: derp-1=cr0s,wr0s
    magicsock.go:635: [v1] magicsock: starting endpoint update (set-private-key)
    derphttp_client.go:287: derphttp.Client.Connect: connecting to derp-1 (test)
    magicsock_test.go:318: conn2 endpoints update
    magicsock.go:941: magicsock: disco key = d:6c98dc2cd023901c
    magicsock.go:2247: [v1] magicsock: got updated network map; 1 peers
    magicsock.go:2277: magicsock: created endpoint key=[bGiA7]: disco=d:6c98dc2cd023901c; derp=1(test) aip=1.0.0.2 ep=127.0.0.1:53545 
    wglog.go:81: [v2] UAPI: Updating private key
    wglog.go:81: [v2] [bGiA7] - Starting
    wglog.go:81: [v2] [bGiA7] - UAPI: Created
    wglog.go:81: [v2] [bGiA7] - UAPI: Updating endpoint
    wglog.go:81: [v2] [bGiA7] - UAPI: Removing all allowedips
    wglog.go:81: [v2] [bGiA7] - UAPI: Adding allowedip
    magicsock.go:941: magicsock: disco key = d:bbbce5f7895c22fd
    magicsock.go:2247: [v1] magicsock: got updated network map; 1 peers
    magicsock.go:2277: magicsock: created endpoint key=[GQujI]: disco=d:92ef406bea80fe55; derp=1(test) aip=1.0.0.1 
    wglog.go:81: [v2] UAPI: Updating private key
    wglog.go:81: [v2] [GQujI] - Starting
    wglog.go:81: [v2] [GQujI] - UAPI: Created
    wglog.go:81: [v2] [GQujI] - UAPI: Updating endpoint
    wglog.go:81: [v2] [GQujI] - UAPI: Removing all allowedips
    wglog.go:81: [v2] [GQujI] - UAPI: Adding allowedip
    magicsock_test.go:318: conn1 endpoints update
    magicsock.go:2247: [v1] magicsock: got updated network map; 1 peers
    magicsock.go:635: [v1] magicsock: starting endpoint update (non-zero-peers)
    wglog.go:81: [v2] [bGiA7] - Sending handshake initiation
    magicsock.go:3586: [v1] magicsock: disco: send, starting discovery for [bGiA7] (d:6c98dc2cd023901c)
    magicsock.go:1431: [v1] magicsock: derp route for [bGiA7] set to derp-1 (shared home)
    magicsock.go:1536: magicsock: derp-1 connected; connGen=1
    magicsock.go:2643: [v1] magicsock: ReSTUN: endpoint update active, need another later ("non-zero-peers")
    magicsock_test.go:318: conn1 endpoints update
    wglog.go:81: [v2] [GQujI] - Received handshake initiation
    wglog.go:81: [v2] [GQujI] - Sending handshake response
    magicsock.go:3586: [v1] magicsock: disco: send, starting discovery for [GQujI] (d:92ef406bea80fe55)
    magicsock.go:1431: [v1] magicsock: derp route for [GQujI] set to derp-1 (shared home)
    magicsock.go:1774: [v1] magicsock: disco: d:6c98dc2cd023901c->d:92ef406bea80fe55 ([GQujI], derp-1) sent call-me-maybe
    magicsock.go:1774: [v1] magicsock: disco: d:6c98dc2cd023901c->d:92ef406bea80fe55 ([GQujI], 127.0.0.1:53541) sent ping tx=3e5faace3127
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53545 derp=1 derpdist=1v4:1ms
    magicsock.go:635: [v1] magicsock: starting endpoint update (non-zero-peers)
    magicsock.go:2247: [v1] magicsock: got updated network map; 1 peers
    magicsock.go:3615: [v1] magicsock: disco: node [GQujI] changed from discokey discokey:92ef406bea80fe555ddab78aa709b5d20b2308e73110f8ba89d433fce36cd961 to discokey:bbbce5f7895c22fd68c0ecf0d669fce74b85c781eda9f73f525de32b7cfded2e
    magicsock.go:1774: [v1] magicsock: disco: d:bbbce5f7895c22fd->d:6c98dc2cd023901c ([bGiA7], derp-1) sent call-me-maybe
    magicsock.go:1774: [v1] magicsock: disco: d:bbbce5f7895c22fd->d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545) sent ping tx=5c556e220b4a
    magicsock.go:2016: [v1] magicsock: disco: d:6c98dc2cd023901c<-d:bbbce5f7895c22fd ([GQujI], 127.0.0.1:53541)  got ping tx=5c556e220b4a
    wglog.go:81: [v2] [bGiA7] - Received handshake response
    magicsock.go:3741: [v1] magicsock: disco: d:bbbce5f7895c22fd<-d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545)  got pong tx=5c556e220b4a latency=6ms pong.src=127.0.0.1:53541
    magicsock.go:3759: magicsock: disco: node [bGiA7] d:6c98dc2cd023901c now using 127.0.0.1:53545
    magicsock.go:1914: [v1] magicsock: disco: d:6c98dc2cd023901c<-d:bbbce5f7895c22fd ([GQujI], derp-1)  got call-me-maybe, 1 endpoints
    magicsock.go:1774: [v1] magicsock: disco: d:6c98dc2cd023901c->d:bbbce5f7895c22fd ([GQujI], 127.0.0.1:53541) sent ping tx=d71f2a5e22f3
    magicsock.go:2016: [v1] magicsock: disco: d:bbbce5f7895c22fd<-d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545)  got ping tx=d71f2a5e22f3
    magicsock_test.go:962: direct link [GQujI]->[bGiA7] found with addr 127.0.0.1:53545
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:900: sending ping stream from [GQujI] (1.0.0.1) to [bGiA7] (1.0.0.2)
    magicsock.go:3741: [v1] magicsock: disco: d:6c98dc2cd023901c<-d:bbbce5f7895c22fd ([GQujI], 127.0.0.1:53541)  got pong tx=d71f2a5e22f3 latency=2ms pong.src=127.0.0.1:53545
    magicsock.go:3759: magicsock: disco: node [GQujI] d:bbbce5f7895c22fd now using 127.0.0.1:53541
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53541 derp=1 derpdist=
    logger.go:31: netcheck: [v1] report: udp=true v6=false mapvarydest= hair=false portmap=? v4a=127.0.0.1:53545 derp=1 derpdist=
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    filter.go:256: Accept: ICMPv4{1.0.0.1:0 > 1.0.0.2:0} 32 ok out
    magicsock_test.go:966: no direct path [bGiA7]->[GQujI] yet, addrs []
    wglog.go:81: [v2] [GQujI] - Sending keepalive packet
    magicsock.go:3586: [v1] magicsock: disco: send, starting discovery for [GQujI] (d:bbbce5f7895c22fd)
    magicsock.go:1774: [v1] magicsock: disco: d:6c98dc2cd023901c->d:bbbce5f7895c22fd ([GQujI], derp-1) sent call-me-maybe
    wglog.go:81: [v2] [bGiA7] - Receiving keepalive packet
    magicsock.go:1774: [v1] magicsock: disco: d:6c98dc2cd023901c->d:bbbce5f7895c22fd ([GQujI], 127.0.0.1:53541) sent ping tx=951dc57c7eef
    magicsock.go:2016: [v1] magicsock: disco: d:bbbce5f7895c22fd<-d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545)  got ping tx=951dc57c7eef
    magicsock.go:1914: [v1] magicsock: disco: d:bbbce5f7895c22fd<-d:6c98dc2cd023901c ([bGiA7], derp-1)  got call-me-maybe, 1 endpoints
    magicsock.go:1774: [v1] magicsock: disco: d:bbbce5f7895c22fd->d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545) sent ping tx=5577b834e9e7
    magicsock.go:3741: [v1] magicsock: disco: d:6c98dc2cd023901c<-d:bbbce5f7895c22fd ([GQujI], 127.0.0.1:53541)  got pong tx=951dc57c7eef latency=0s pong.src=127.0.0.1:53545
    magicsock.go:3741: [v1] magicsock: disco: d:bbbce5f7895c22fd<-d:6c98dc2cd023901c ([bGiA7], 127.0.0.1:53545)  got pong tx=5577b834e9e7 latency=0s pong.src=127.0.0.1:53541
    magicsock_test.go:970: magicsock did not find a direct path from [bGiA7] to [GQujI]
    wglog.go:81: [v2] Device closing
    wglog.go:81: [v2] Routine: receive incoming v4 - stopped
    wglog.go:81: [v2] Routine: receive incoming receiveDERP - stopped
    wglog.go:81: [v2] Routine: receive incoming v6 - stopped
    wglog.go:81: [v2] [GQujI] - Stopping
    wglog.go:81: [v2] Device closed
    magicsock.go:3914: [v1] magicsock: doing cleanup for discovery key d:bbbce5f7895c22fd
    magicsock.go:2346: magicsock: closing connection to derp-1 (conn-close), age 10s
    magicsock.go:2356: magicsock: 0 active derp conns
    wglog.go:81: [v2] Device closing
    wglog.go:81: [v2] Routine: receive incoming v4 - stopped
    logger.go:31: derp client 127.0.0.1:50214/6e6f64656b65793a36633638383065633235636364623337313163353631356262336562643834343964643830616563393532353962623166623565653934343030343931373766: read EOF
    wglog.go:81: [v2] Routine: receive incoming receiveDERP - stopped
    wglog.go:81: [v2] Routine: receive incoming v6 - stopped
    wglog.go:81: [v2] [bGiA7] - Stopping
    wglog.go:81: [v2] Device closed
    magicsock.go:3914: [v1] magicsock: doing cleanup for discovery key d:6c98dc2cd023901c
    magicsock.go:2346: magicsock: closing connection to derp-1 (conn-close), age 10s
    magicsock.go:2356: magicsock: 0 active derp conns
    logger.go:31: derp client 127.0.0.1:50214/6e6f64656b65793a36633638383065633235636364623337313163353631356262336562643834343964643830616563393532353962623166623565653934343030343931373766: removing connection
    logger.go:31: derp client 127.0.0.1:50213/6e6f64656b65793a31393062613332313735313933333734396139613965353734333036393036323861666630336534633362363336366536353833303433343239396466363665: read EOF
    logger.go:31: derp client 127.0.0.1:50213/6e6f64656b65793a31393062613332313735313933333734396139613965353734333036393036323861666630336534633362363336366536353833303433343239396466363665: removing connection
    stuntest.go:64: STUN server shutdown
FAIL
```

### Steps to reproduce

_No response_

### Are there any recent changes that introduced the issue?

_No response_

### OS

_No response_

### OS version

_No response_

### Tailscale version

_No response_

### Bug report

_No response_

--- a/wgengine/magicsock/magicsock_test.go
+++ b/wgengine/magicsock/magicsock_test.go
@@ -955,8 +955,9 @@ func testActiveDiscovery(t *testing.T, d *devices) {
 
 func mustDirect(t *testing.T, logf logger.Logf, m1, m2 *magicStack) {
 	lastLog := time.Now().Add(-time.Minute)
-	// See https://github.com/tailscale/tailscale/issues/654 for a discussion of this deadline.
-	for deadline := time.Now().Add(10 * time.Second); time.Now().Before(deadline); time.Sleep(10 * time.Millisecond) {
+	// See https://github.com/tailscale/tailscale/issues/654
+	// and https://github.com/tailscale/tailscale/issues/3247 for discussions of this deadline.
+	for deadline := time.Now().Add(30 * time.Second); time.Now().Before(deadline); time.Sleep(10 * time.Millisecond) {
 		pst := m1.Status().Peer[m2.Public()]
 		if pst.CurAddr != "" {
 			logf("direct link %s->%s found with addr %s", m1, m2, pst.CurAddr)

